### PR TITLE
Upgrade middleman [WIP]

### PR DESCRIPTION
**Pull request by JonRowe**
Todo:

- [ ] fix assets
- [ ] restore deploy
- [ ]  restore or just hardcode favicon

### CODE REVIEW

File: config.rb [hunk 1]
```diff
@@ -1,6 +1,11 @@
 $LOAD_PATH << File.expand_path('../lib', __FILE__)
 require 'patches/uri'
 
+# Assets
+# Framework requires set themselves up
+require 'bourbon'
+require 'neat'
```
Comment (pirj) [inline config.rb:7]: `Bundler.require` maybe?

File: config.rb [hunk 2]
```diff
@@ -114,9 +120,9 @@ def deploy_to target
 require 'rspec_info/helpers'
 helpers RSpecInfo::Helpers
 
-RSpecInfo::Helpers.rspec_documentation_latest(source_dir).each do |gem_name, version|
-  Dir.glob(File.join(source_dir, "/documentation/#{version}/#{gem_name}/**/*")).select { |f| File.file? f }.each do |f|
-    relative_path = Pathname.new(f).relative_path_from(Pathname.new(source_dir))
+RSpecInfo::Helpers.rspec_documentation_latest(app.source_dir).each do |gem_name, version|
+  Dir.glob(File.join(app.source_dir, "/documentation/#{version}/#{gem_name}/**/*")).select { |f| File.file? f }.each do |f|
```
Comment (pirj) [inline config.rb:108]: I ended up with `**/*.html`, otherwise it was replacing CSS and JS with some nonsense.

File: Gemfile [hunk 3]
```diff
@@ -1,12 +1,14 @@
 source 'https://rubygems.org'
 
-gem "middleman", "~> 3.3.6"
+gem "middleman", "~> 4.0.0"
 gem "middleman-syntax"
 gem "middleman-blog"
 gem "middleman-favicon-maker"
 gem "middleman-deploy"
```
Comment (pirj) [inline Gemfile]: `middleman-deploy` won't work, can be dropped.

**Review by pirj — COMMENTED** [review]
Is this needed? They don't seem to use any SCSS syntax (AFAIR `@keyframes` isa native CSS) `asciinema-player.css` -> `asciinema-player.scss`?

Side observation: a good part of Asciinema CSS can be dropped, as it contains styles for different themes, and we only use one.

The original is in LESS.

Random notes.

The idea with submodules for deployment should work. I didn't play much around it, but was able to create some mess inside `build` directory.

WDYT if I send a PR to replace `middleman-deploy` with submodules first, and we apply this PR with Middleman 3->4 upgrade after with less pain?

**Comment by JonRowe** [comment]
> The idea with submodules for deployment should work. I didn't play much around it, but was able to create some mess inside build directory.
> 
> WDYT if I send a PR to replace middleman-deploy with submodules first, and we apply this PR with Middleman 3->4 upgrade after with less pain?

I mean thats an orthogonal approach, so it doesn't hurt to try, (replacing deploy before upgrading), but remember the source of the master branch is what github sees and deploys, so I'm sceptical that submodules can work.

File: config.rb (hunk 1 above)
Comment (JonRowe) [reply config.rb:7]: I'm only replacing what is missing from the asset pipeline, I'm not sure requiring everything is a good idea.

File: config.rb (hunk 2 above)
Comment (JonRowe) [reply config.rb:108]: I mean this is supposed to replace css and js...

**Comment by JonRowe** [comment]
> Is this needed? They don't seem to use any SCSS syntax (AFAIR @keyframes isa native CSS) asciinema-player.css -> asciinema-player.scss?
> 
> Side observation: a good part of Asciinema CSS can be dropped, as it contains styles for different themes, and we only use one.

Its the as bundled file, the file was renamed as an attempt to get sass rendering to work (I think it's bundled?) if its not bundled it can probably be left alone.

**Comment by JonRowe** [comment]
I've been able to mostly get HTML rendering working, whats needed mostly is a replacement/restore for sprockets

**Comment by pirj** [comment]
Got a response from support:

----
The CNAME in the repo is what’s used to direct traffic to the right place. What Custom Domain settings does is to create the CNAME file in the repo in an automated fashion. The CNAME isn’t being stored anywhere else.

>    what is the best way to achieve a "git push"-style deployment to both the stable main site (with a canonical domain name) and a Staging site (with no canonical name) from a single repository?

You still need a way to prevent the CNAME file from being pushed to staging.

I'm sorry I could not be of much help here, you may have to find a third-party tool similar "Middleman" or otherwise reach out to "Middleman" support team to check your options after their recent update.
----

I guess we'll have to use something like this to push to staging:
```
git rm docs/CNAME
git commit docs -m 'rm CNAME'
git push staging HEAD:source
git reset --hard HEAD~1
```
This is what our previous `middleman deploy` was doing anyway, right?

**Comment by JonRowe** [comment]
Previously Github had `master` hardcoded as the page branch, so `middleman deploy` built the site, checked out master, copied the files and commited them, and thus it could ignore files when copying.

**Comment by JonRowe** [comment]
Maybe your submodule approach is cleaner here, because build will output to a directory, and we can commit and push to the submodules (preferably with a helper command) similar to how the old middleware worked.

That way a commit deleting CNAME will stay in the history and as long as we never change CNAME it'll persist.

**Comment by pirj** [comment]
We're already on Middleman 4.x, done in #163
@JonRowe Do you think there's anything useful left in this PR, or it can be closed?

**Comment by JonRowe** [comment]
Nope! Closed :)

**Pull request closed without merging**